File: flattenBinaryTreeToLinkedList.py
# problem: https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
# Runtime: 36 ms, faster than 74.13% of Python3 online submissions for Flatten Binary Tree to Linked List.
# Memory Usage: 15.3 MB, less than 12.74% of Python3 online submissions for Flatten Binary Tree to Linked List.


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def traverse(self, node):
        if not node:
            return node, None
        elif not node.left and not node.right:
            return node, node
        l, lt = self.traverse(node.left)
        r, rt = self.traverse(node.right)
        if l and lt:
            node.right = l
            lt.right = r
        else:
            node.right = r
        if not rt:
            rt = lt
        node.left = None
        return node, rt

    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.traverse(root)

File: flattenBinaryTreeToLinkedList.cpp
// problem:https://leetcode.com/problems/flatten-binary-tree-to-linked-list/
// Runtime: 7 ms, faster than 81.17% of C++ online submissions for Flatten Binary Tree to Linked List.
// Memory Usage: 12.6 MB, less than 96.54% of C++ online submissions for Flatten Binary Tree to Linked List.

#include <bits/stdc++.h> 

using namespace std;


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* traverse(TreeNode* root) {
        if (!root) return root;
        
        TreeNode* cur = traverse(root -> left);
        TreeNode* tmp = root -> right;
        while (cur && cur -> right) cur = cur -> right;
        if (cur) {
            root -> right = root -> left;
            cur -> right = traverse(tmp);
        } else {
            root -> right = traverse(tmp);
        }
        root -> left = nullptr;
        return root;
    }
    
    void flatten(TreeNode* root) {
        traverse(root);
    }
};
